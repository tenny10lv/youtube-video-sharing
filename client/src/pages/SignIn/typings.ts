export enum Field {
  EMAIL = "email",
  PASSWORD = "password",
}

export type FormField = {
  [Field.EMAIL]: string;
  [Field.PASSWORD]: string;
};
