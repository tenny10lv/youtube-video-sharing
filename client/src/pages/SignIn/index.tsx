import React, { useCallback, useState } from "react";
import Button from "@mui/material/Button";
import CssBaseline from "@mui/material/CssBaseline";
import TextField from "@mui/material/TextField";
import FormControlLabel from "@mui/material/FormControlLabel";
import Checkbox from "@mui/material/Checkbox";
import Link from "@mui/material/Link";
import Grid from "@mui/material/Grid";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import Container from "@mui/material/Container";
import { useAuth } from "../../context/AuthContext";
import { useNavigate } from "react-router-dom";
import { Alert } from "@mui/material";
import { useFormik } from "formik";
import * as Yup from "yup";
import { Field, FormField } from "./typings";

function Copyright(props: any) {
  return (
    <Typography
      variant="body2"
      color="text.secondary"
      align="center"
      {...props}
    >
      {"Copyright Â© "}
      <Link color="inherit" href="https://mui.com/">
        Remitano
      </Link>{" "}
      {new Date().getFullYear()}
      {"."}
    </Typography>
  );
}

const signInSchema = Yup.object().shape({
  email: Yup.string().email("Invalid email").required("Required"),
  password: Yup.string().required("Required"),
});

const SignIn = () => {
  const auth = useAuth();
  const navigate = useNavigate();

  const [message, setMessage] = useState<string>();
  const [loading, setLoading] = useState(false);

  const onSubmit = async (values: FormField) => {
    setLoading(true);
    const user = await auth?.logIn(values[Field.EMAIL], values[Field.PASSWORD]);
    setLoading(false);
    if (user) {
      navigate("/");
    } else {
      setMessage("Email or password is not correct");
    }
  };

  const formik = useFormik<FormField>({
    initialValues: {
      [Field.EMAIL]: "",
      [Field.PASSWORD]: "",
    },
    validationSchema: signInSchema,
    onSubmit,
  });

  const getError = useCallback(
    (field: Field) => {
      return formik.touched[field] && formik.errors[field];
    },
    [formik]
  );

  const handleSubmit = (event: any) => {
    event.preventDefault();
    formik.handleSubmit();
  };

  return (
    <Container component="main" maxWidth="xs">
      <CssBaseline />
      <Box
        sx={{
          marginTop: 8,
          display: "flex",
          flexDirection: "column",
          alignItems: "center",
        }}
      >
        <Typography component="h1" variant="h5">
          Sign In
        </Typography>
        {message && (
          <Alert severity="error" sx={{ width: "100%", margin: "20px 0" }}>
            {message}
          </Alert>
        )}
        <Box component="form" onSubmit={handleSubmit} noValidate sx={{ mt: 1 }}>
          <TextField
            name={Field.EMAIL}
            label="Email Address *"
            margin="normal"
            fullWidth
            error={!!getError(Field.EMAIL)}
            helperText={getError(Field.EMAIL)}
            value={formik.values.email}
            onChange={formik.handleChange}
          />
          <TextField
            name={Field.PASSWORD}
            label="Password *"
            margin="normal"
            fullWidth
            type="password"
            error={!!getError(Field.PASSWORD)}
            helperText={getError(Field.PASSWORD)}
            value={formik.values.password}
            onChange={formik.handleChange}
          />
          <FormControlLabel
            control={<Checkbox value="remember" color="primary" />}
            label="Remember me"
          />
          <Button
            type="submit"
            fullWidth
            variant="contained"
            sx={{ mt: 3, mb: 2 }}
            disabled={loading}
          >
            Sign In
          </Button>
          <Grid container>
            <Grid item xs>
              <Link href="#" variant="body2">
                Forgot password?
              </Link>
            </Grid>
            <Grid item>
              <Link href="/signup" variant="body2">
                {"Don't have an account? Sign Up"}
              </Link>
            </Grid>
          </Grid>
        </Box>
      </Box>
      <Copyright sx={{ mt: 8, mb: 4 }} />
    </Container>
  );
};

export default SignIn;
