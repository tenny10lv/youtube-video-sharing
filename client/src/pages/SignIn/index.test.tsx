import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import * as authContext from "../../context/AuthContext";
import SignIn from "./";

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...(jest.requireActual("react-router-dom") as any),
  useNavigate: () => mockedUsedNavigate,
}));

describe("signin-page", () => {
  test("Should match snapshot", () => {
    const { container } = render(<SignIn />);
    expect(container).toMatchSnapshot();
  });

  test("render signin page", async () => {
    render(<SignIn />);

    const emailInput = screen.getByLabelText("Email Address *") as any;
    expect(emailInput).toBeInTheDocument();

    const passwordInput = screen.getByLabelText("Password *") as any;
    expect(passwordInput).toBeInTheDocument();

    const signInButton = screen.getByRole("button", {
      name: "Sign In",
    });
    expect(signInButton).toBeInTheDocument();
  });

  test("signin page - validate input", async () => {
    render(<SignIn />);

    const signInButton = screen.getByRole("button", {
      name: "Sign In",
    });

    // empty inputs
    fireEvent.click(signInButton);
    const requiredTexts = await screen.findAllByText("Required");
    expect(requiredTexts.length).toEqual(2);
  });

  test("signin page - signed in successfully", async () => {
    jest.spyOn(authContext, "useAuth").mockReturnValue({
      logIn: () => ({ id: 1, username: "test@yopmail.com" }),
    } as any);

    render(<SignIn />);

    const emailInput = screen.getByLabelText("Email Address *") as any;
    fireEvent.change(emailInput, {
      target: { value: "test@yopmail.com" },
    });

    const passwordInput = screen.getByLabelText("Password *") as any;
    fireEvent.change(passwordInput, {
      target: { value: "123456" },
    });

    const signInButton = screen.getByRole("button", {
      name: "Sign In",
    });
    fireEvent.click(signInButton);

    await waitFor(() => expect(mockedUsedNavigate).toHaveBeenCalledWith("/"));
  });

  test("signin page - signed in failed", async () => {
    // const navigate = useNavigate();
    jest.spyOn(authContext, "useAuth").mockReturnValue({
      logIn: () => undefined,
    } as any);

    render(<SignIn />);

    const emailInput = screen.getByLabelText("Email Address *") as any;
    fireEvent.change(emailInput, {
      target: { value: "test@yopmail.com" },
    });

    const passwordInput = screen.getByLabelText("Password *") as any;
    fireEvent.change(passwordInput, {
      target: { value: "123456" },
    });

    const signInButton = screen.getByRole("button", {
      name: "Sign In",
    });
    fireEvent.click(signInButton);

    expect(
      await screen.findByText("Email or password is not correct")
    ).toBeVisible();
  });
});
