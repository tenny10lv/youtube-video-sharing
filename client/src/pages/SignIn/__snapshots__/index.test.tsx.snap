// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signin-page Should match snapshot 1`] = `
<div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-binzgt"
    >
      <h1
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Sign In
      </h1>
      <form
        class="MuiBox-root css-164r41r"
        novalidate=""
      >
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Email Address *
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              name="email"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Email Address *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Password *
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Password *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
              value="remember"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Remember me
          </span>
        </label>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-1vhaqj4-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Sign In
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
          >
            <a
              class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineAlways css-101ca9i-MuiTypography-root-MuiLink-root"
              href="#"
            >
              Forgot password?
            </a>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <a
              class="MuiTypography-root MuiTypography-body2 MuiLink-root MuiLink-underlineAlways css-101ca9i-MuiTypography-root-MuiLink-root"
              href="/signup"
            >
              Don't have an account? Sign Up
            </a>
          </div>
        </div>
      </form>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter css-ijul54-MuiTypography-root"
    >
      Copyright Â© 
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1cghntr-MuiTypography-root-MuiLink-root"
        href="https://mui.com/"
      >
        Remitano
      </a>
       
      2023
      .
    </p>
  </main>
</div>
`;
